use-cases/GPU-passthrough-and-Kata.md

```diff
@@ -0,0 +1,270 @@
+```
+
+Make sure you are using the `pc` machine type by verifying `machine_type = "pc"` is 
+set in the configuration.toml.
+
+## Build Kata Containers kernel with GPU support
+
+The default guest kernel installed with Kata Containers does not provide GPU support.
+To use an Intel GPU with Kata Containers, you need to build a kernel with the necessary
+GPU support.
+
+The following i915 kernel config options need to be enabled:
+```
+CONFIG_DRM=y
+CONFIG_DRM_I915=y
+CONFIG_DRM_I915_USERPTR=y
+```
+
+Build the Kata Containers kernel with the previous config options, using the instructions
+described in [Building Kata Containers kernel](https://github.com/kata-containers/packaging/tree/master/kernel).
+For further details on building and installing guest kernels, see [the developer guide](https://github.com/kata-containers/documentation/blob/master/Developer-Guide.md#install-guest-kernel-images).
+
+## GVT-d with Kata Containers
+
+Use the following steps to pass an Intel Graphics device in GVT-d mode with Kata:
+
+1. Find the Bus-Device-Function (BDF) for GPU device:
+
+   ```
+   $ sudo lspci -nn -D | grep Graphics
+     0000:00:02.0 VGA compatible controller [0300]: Intel Corporation Broadwell-U Integrated Graphics [8086:1616] (rev 09)
+   ```
+
+   Run the previous command to determine the BDF for the GPU device on host.<br/>
+   From the previous output, PCI addres "0000:00:02.0" is assigned to the hardware GPU device.<br/>
+   This BDF is used later to unbind the GPU device from the host.<br/>
+   "8086 1616" is the device ID of the hardware GPU device. It is used later to
+   rebind the GPU device to `vfio-pci` driver.
+
+2. Find the IOMMU group for the GPU device:
+
+   ```
+   $ BDF="0000:00:02.0"
+   $ readlink -e /sys/bus/pci/devices/$BDF/iommu_group
+   ../../../../kernel/iommu_groups/1
+   ```
+
+   The previous output shows that the GPU belongs to IOMMU group 1.
+
+3. Unbind the GPU:
+
+   ```
+   $ echo $BDF | sudo tee /sys/bus/pci/devices/$BDF/driver/unbind
+   ```
+
+4. Bind the GPU to the `vfio-pci` device driver:
+
+   ```
+   $ sudo modprobe vfio-pci
+   $ echo 8086 1616 | sudo tee /sys/bus/pci/drivers/vfio-pci/new_id
+   $ echo $BDF | sudo tee --append /sys/bus/pci/drivers/vfio-pci/bind
+   ```
+
+   After you run the previous commands, the GPU is  bound to `vfio-pci` driver.<br/>
+   A new directory with the IOMMU group number is created under `/dev/vfio`:
+
+   ```
+   $ ls -l /dev/vfio
+     total 0
+     crw------- 1   root root  241,   0 May 18 15:38 1
+     crw-rw-rw- 1 root root  10, 196 May 18 15:37 vfio
+   ```
+
+5. Start a Kata container with GPU device:
+
+   ```
+   $ sudo docker run -it --runtime=kata-runtime --rm --device /dev/vfio/1 -v /dev:/dev debian /bin/bash
+   ```
+
+   Run `lspci` within the container to verify the GPU device is seen in the list of 
+   the PCI devices.
+
+   ```
+   $ lspci -nn -D
```

Comment:
@amshinde can you add -v and get the details here. The item of interest is the PCI BAR programming?

Reply:
@amshinde we hotplug the device behind the pci bridge right? Does our PCI bridge have enough MMIO space allocated to it?  What can the user do if the graphics device needs a larger amount of memory?

Reply:
@mcastelino Yes, thats why I mentioned above to turn on the configuration `hotplug_vfio_on_root_bus` for pc platform. With that the GPU is hotplugged on the main bus. Its not possible to change the MMIO space allocated for a pci bridge. 
You can do that for a pcie bridge, but I havent tested this with q35 yet. Will add instructions for q35 in a follow up PR once I have tested it with a GPU.